internal/mutator/namespace_included_label.go

```diff
@@ -0,0 +1,77 @@
+package mutator
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	api "sigs.k8s.io/hierarchical-namespaces/api/v1alpha2"
+	"sigs.k8s.io/hierarchical-namespaces/internal/config"
+)
+
+// NamespaceIncludedLabelServingPath is where the mutator will run. Must be kept
+// in sync with the kubebuilder markers below.
+const (
+	NamespaceIncludedLabelServingPath = "/mutate-namespace-included-label"
+)
+
+// Note: the mutating webhook FAILS OPEN. This means that if the webhook goes
+// down, all further changes are allowed. (An empty line has to be kept below
+// the kubebuilder marker for the controller-gen to generate manifests.)
+//
+// +kubebuilder:webhook:admissionReviewVersions=v1;v1beta1,path=/mutate-namespace-included-label,mutating=true,failurePolicy=ignore,groups="",resources=namespaces,sideEffects=None,verbs=create;update,versions=v1,name=namespacelabel.hnc.x-k8s.io
+
+type NamespaceIncludedLabel struct {
+	Log     logr.Logger
+	decoder *admission.Decoder
+}
+
+// Handle implements the mutating webhook.
+func (n *NamespaceIncludedLabel) Handle(ctx context.Context, req admission.Request) admission.Response {
+	log := n.Log.WithValues("namespace", req.Name)
+	ns := &corev1.Namespace{}
+	err := n.decoder.Decode(req, ns)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	log.V(1).Info(fmt.Sprintf("%+v\n", ns))
+
+	n.handle(log, ns)
+	marshaledNS, err := json.Marshal(ns)
+	if err != nil {
+		return admission.Errored(http.StatusInternalServerError, err)
+	}
+	return admission.PatchResponseFromRaw(req.Object.Raw, marshaledNS)
+}
+
+// handle implements the non-boilerplate logic of this mutator, allowing it to
+// be more easily unit tested (ie without constructing a full admission.Request).
+func (n *NamespaceIncludedLabel) handle(log logr.Logger, ns *corev1.Namespace) {
+	isExcluded := config.ExcludedNamespaces[ns.Name]
+
+	// Ensure the labels map is not nil and remove the included-namespace label if
+	// it exists.
+	if ns.Labels == nil {
+		ns.Labels = map[string]string{}
+	}
+	delete(ns.Labels, api.LabelIncludedNamespace)
```

Comment:
I think it might be worth distinguishing between labels that are _missing_ or _wrong_. If a label is missing, we should add it, but if a label is _wrong_, we should reject it rather than silently change it. E.g. if `kube-system` has `included-namespace`, I think we should just reject that; if another namespace has `included-namespace=cat`, we should reject that as well. What you've written here is certainly _correct_ but I think it would be surprising, since people expect defaults to be set, but not for their explicitly created values to change.

You can either keep that logic in the validating webhook if you like (that's kind of thematically nice) and _only_ add in the missing labels here, or you could reject bad labels here as well. Up to you, but the first option will be a smaller diff.

Note that the validating webhooks always run _after_ the mutating webhooks so you can make the validating webhooks very strict (e.g. the label must exist and must have the correct value).

Reply:
SG. I will only mutate included namespace with missing label to have the label and restore the VWH rules.

BTW in that case, we still need a client in the namespace VWH to get the namespace instance to compare the instance vs the request to deny the request if the labels are **newly**-added on excluded namespace or **newly**-modified on the included namespace, right?

Reply:
Discussed offline - the Request will include the old copy of the namespace inst.